### ISSUE TITLE
Cleaning USDA seed manual data

### ISSUE THREAD

**Issue by DeirdreLoughnan** [lizzieinvancouver/egret#20]
@lizzieinvancouver @wangxm-forest @sandyie @selenashew 
I met with Selena this morning and walked through the scraped data. It looks great and I am very impressed! I took some time to think though our aims and how best to move forward with the cleaning.

@selenashew could you begin by completing the following tasks:
- [x] Add a readme for the scrapeUSDAseedmanual folder explaining what the number folder names mean and where to find key documents
- [x] Double check the scraped tables and update the names of any you think are relevant to our project
- [x] Include in the readme a count of the number of tables scraped for the three data types
- [x] Add a list to the readme for what columns are present in some tables but not included in the [main data file](https://github.com/lizzieinvancouver/egret/blob/main/analyses/scrapeUSDAseedmanual/cleaning/scraped_data_master_spreadsheet.xlsx), it sounds like some are not currently included because they only occur in a few tables. If there are a lot (20+), then skip this task for now.
- [x] Either recreate the main data file, or add a column with the table/page number so we know what table needs to be referenced for each. I think we should keep the current columns you have.

@lizzieinvancouver my first impression is that there is some useful data here, but that the cleaning will require cross referencing each table a little. For example, the table data does not include the genus names (species names are given as "A. rubra" for example), so in some cases it is hard to know the exact species the data is for.  Depending on the number of tables, this could be a good task to divide between several of us.

My proposal is to prioritize the data for the germination_test_conditions, so we can have it clean by the retreat in July. But @lizzieinvancouver @wangxm-forest do you have other data you need prioritized? Once we have a sense of how many tables need to be checked, we can decide how many people should help with this task.

**Comment by lizzieinvancouver** [comment]
> Depending on the number of tables, this could be a good task to divide between several of us.

@DeirdreLoughnan I agree with the above!

> My proposal is to prioritize the data for the germination_test_conditions, so we can have it clean by the retreat in July. 

@DeirdreLoughnan I also agree with this! And from there I could come up with a cascade of priority if you remind where the table types list is.

**Comment by DeirdreLoughnan** [comment]
@selenashew It looks like you have made some progress on getting the information needed above! But how close are you to being finished the five tasks above?

I spoke with @buniwuuu and @ngoj1 about helping to clean this data. We would like to have the germination_test_conditions cleaned by July 8th. So with this in mind, we should prioritize getting task 2 and 5 above done for the germination data at least---with a raw data csv file of the germination_test_condition data saved by the end of this week (Friday June 15th). 

Once we have this new data table, could the three of you divide the number of tables equally and start scripts like we have for the scraped paper data to clean the data? At minimum we need scripts/code that will:

1. Add a column for a. genus name, b. species name that does not include the genus initial, c. Rename the current species column to "species_name" and d. code to clean each ( ie. move the information on source population or seed types, check names are spelt correctly).
4. Add a column for table number and one for page number to make future cross referencing easier
5. Add a column for seed type---some studies seem to specify "dried" vs "fresh"
6. Clean the data to have "NA" instead of "-", but for now we can refrain from averaging across ranges etc

Thank you all for your help on this!

**Comment by lizzieinvancouver** [comment]
> Thank you all for your help on this!

My thanks also! It is exciting that we should have these data soon.

**Comment by selenashew** [comment]
Hi @DeirdreLoughnan,

My apologies for the late reply, I just returned from Seattle! 

As an update, I have completed the original first 3 tasks and have pushed those changes to GitHub (wrote scripts to automate renaming the relevant data tables and to count the total number of data tables per each type, and updated the README). I have also been thinking through on how to tackle the actual amalgamation of all the germination data, and here are my thoughts in regards to the 4 new tasks you've specified:

In general: before diving into these 4 tasks, a lot of manual work is going to be required. The data is incredibly messy across all of the germination tables, in that a lot of the data column names have been arbitrarily shifted or misspelled, and do not line up with the actual data below. Since the shifts and changes are arbitrary and differ from table to table (which makes it near impossible to write a script for since we cannot just code a line that says for ex. "look through cell 1 of column 1 of each csv and append the first word to the new column called 'genus_name'"), I will need to manually set up the master germination spreadsheet with the correct column names, go through all of the tables, and copy and paste the correct data into the spreadsheet to most easily set it up. From there, we can then create scripts to automate other data cleaning tasks such as replacing all of the "-" with "NA" values.

1. Adding a column for the genus names will have to be manually done as we will need to cross-reference with the original PDF.   The other tasks can definitely be automated.
2. This will also have to be manually done as we will need to check the PDF for the exact page number. The scraped data table folders only lists the range of pages that the specific table falls between (eg. between pages 701-810). 
3. Sounds good, I'll make sure to include this column in the master germination spreadsheet! To confirm, the data columns to include are: **Table_Number, Page_Number, Genus_Name, Species_Name, Seed_Source, Seed_Type, Medium, Pretreatment, Cold_Stratification_Days, Daily_Light_Hours, Day_Temp_Celsius, Night_Temp_Celsius, Temp_Unspecified_Time_Of_Day_Celsius, Mean_Light, Mean_Dark, Light_Range, Dark_Range, Treatment_Duration_in_Days, Germination_Time_Days, Total_Germination_Percent, Mean_Germination_Percent**. If there are any columns that you feel are unnecessary, I will happily remove them.
4. We can definitely write a script for this, along with other data cleaning items- such as removing all of the random quotation marks, "I" values, and "+" signs (so 62+ just becomes 62).

In Summary: 
- I will set up a meeting to discuss all of this between the three of us and to divide up the tables to work on
- @buniwuuu, @ngoj1, and I will then do manual work just to compile the correct data into the correct locations on the master germination spreadsheet
- We will then write a script for other data cleaning tasks

Would it be possible to **extend the June 15 deadline to June 20 instead since I just got back?** This will give me more time to schedule a meeting between Justin, Britney, and I to discuss this task. If not, I will get all of the manual work completed and ready by myself for June 15, and then I will hand off the scriptwriting tasks to Justin and Britney after.

**Comment by lizzieinvancouver** [comment]
> Would it be possible to **extend the June 15 deadline to June 20 instead since I just got back?** This will give me more time to schedule a meeting between Justin, Britney, and I to discuss this task. If not, I will get all of the manual work completed and ready by myself for June 15, and then I will hand off the scriptwriting tasks to Justin and Britney after.

@selenashew Thanks for your work on this! Unfortunately, we are REALLY running out of time to do this, so I don't think an extension of almost a week is a good idea. I would ask you try to finish this sooner.

**Comment by selenashew** [comment]
Hi everyone,

As an update, **I have finished compiling the master germination data spreadsheet, which can be found under "egret\analyses\scrapeUSDAseedmanual\cleaning\germination_master_spreadsheet.xlsx"**. This spreadsheet is much cleaner compared to the first version. 

There were several columns that were not included as they only appeared once or twice across all of the scraped germination data tables. These columns are:

- soundness
- purity
- elevation
- seed fill
- seed age
- viability
- germination speed
- germination range
- (number of) tests
- germinative capacity range
- time to 50% germination
- seeds treated (entire column was blank)

Please let me know if any of these columns may be needed and I can add them in.

On Monday, Justin and I will meet up to discuss the project and to write a script to finalize data cleaning for the master germination spreadsheet. Our script will aim to:

- Remove all misc. symbols (I, -, #, $)
- Add NA values to cells with missing values
- Remove all of the quotation marks
- Remove any duplicate rows
- Remove (+) signs in the species_name column
- For the species_name column, fill missing values with the cell value above it (as the species repeats)
- Remove any completely empty columns
- Add unique IDs to each table (the current table numbers repeat from 1-10)
- Address any other weird issues we may spot on closer review

The data should be ready to go right afterwards! :)

We will then move onto compiling and cleaning the phenology data tables. I have gone through them and they appear to be in a much better raw shape (AKA scraped in a much more standardized manner) compared to the germination tables, so we should be able to write more scripts to speed up the compilation/cleaning process specifically for them.

**Comment by lizzieinvancouver** [comment]
> There were several columns that were not included as they only appeared once or twice across all of the scraped germination data tables. These columns are:
> 
> * soundness
> * purity
> * elevation
> * seed fill
> * seed age
> * viability
> * germination speed
> * germination range
> * (number of) tests
> * germinative capacity range
> * time to 50% germination
> * seeds treated (entire column was blank)

@selenashew Amazing work -- thank you!

As for the above, we should get the data in:
* germination speed
* time to 50% germination
* elevation
* seed age
And can someone double-check the original table to make sure `seeds treated` really is blank? Otherwise that is likely a column we would want (but not if empty!).

**Comment by buniwuuu** [comment]
Hi @selenashew !
I'm currently working through scrapeUSDAseedmanual/cleaning/phenology_data_preparation_script.R and have some questions:

1. Do all the tables have different formats (I assume yes)? In the test folder, the column names can occupy multiple rows (for example, germination/rate/(%) each have their own cell), so we might have to keep rows above the start of real data to be able to know the column names. However, that leaves the dataframe messy. 
2. Will we manually select tables that will be cleaned with this code? 
3. Should we write the files into the output folder Instead of updating the input files?

Here are the things I fixed that seem to successfully add genus and pdf table number columns (still looking pretty rough due to the formatting problem above):
1. In `read_csv()` added `col.name = FALSE` so "table#" can be detected when it's in the first row
2. Change `str_detect()` to `str_match()` because "Table ([0-9]+)-(.*?)(,|$)" has variable length that isn't allowed in`str_detect()`

**Comment by selenashew** [comment]
Hi everyone,

Here are the current updates:

1) I have added in the columns for:
-germination speed
-time to 50% germination
-elevation
-seed age
-I checked and turns out the 'seed_treated' column was actually the 'pretreatment' column, and have manually added in those data values accordingly

2) Justin and I have had two meetings to work on the project together. Justin is currently finishing up the script to complete cleaning the germination data, and it should be ready to go by Wednesday.

3) Last night, I coded up the rough draft of a script to automate preparing all of the phenology data for merging, and have now passed that onto to Britany to review & edit. I just had a meeting with Britany to walk through the prep/cleaning of the phenology data, and have answered her questions in person. As a record of responses to her questions:

1-2. The germination data was unfortunately scraped quite terribly, and as you've noticed, are all extremely messy. Fortunately, the phenology data was all scraped quite cleanly, and the formats are all standardized. I have pulled all of the phenology data and have made copies in 2 different folders- the 'copy of all phenology data' folder contains all of the scraped phenology data that we are going to prep, merge, and clean, and there is a test folder to play around with while working on our phenology scripts.
3. That's a great point- yes, let's save the outputs of the phenology prep script to the output folder!
4. Thank you so much Britany for your help and for diving right back into the chaos :)

**Comment by lizzieinvancouver** [comment]
@selenashew Thanks for all your work on this!

**Comment by ngoj1** [comment]
Germination data is almost fully cleaned up now, I just had two questions:
1) How should we categorize the pretreatment categories? To my knowledge mechanical, nicking, and abrasion are all physical/mechanical forms of scarification, so should we leave them as they are now or change them to "mechanical"? Here's a photo of all the different kinds of pretreatments:
![image](https://github.com/lizzieinvancouver/egret/assets/87724921/27c13838-a661-4c7e-99b8-d2a9636f6dba)

2) I've tried my best to fix any mistakes in the species names that stood out as obviously wrong, but I don't know all the species names in the world and so was wondering how to best correct any other mistakes I may have missed. For example, Rubus idaeus was sometimes spelled Rubus idaeust with the "t". I remember for EGRET we compared against a plant database (from Kew? Plants of the World?) so if anyone could give some pointers as to how to approach this, please let me know!

**Comment by DeirdreLoughnan** [comment]
Thanks @ngoj1 and @selenashew for your help with this!

> 1. How should we categorize the pretreatment categories? To my knowledge mechanical, nicking, and abrasion are all physical/mechanical forms of scarification, so should we leave them as they are now or change them to "mechanical"? 

It would be ideal if we could replicate what was done for the other egret dataset when possible. Could you clean the data to create new columns based on these pretreatments. Scanning the list, I would create columns for: 

1. Seed type---which would include fresh vs all the ones from feces
2. Chill and chill_duration---here we could create two new columns, one for whether seeds were stratified (Y/N) and a duration column that includes the length but converted to days, since that is the unit we use for the other dataset. 
3. Scarification---similar to the main data can we create two cleaned columns: scarifType---which was very specific i.e. "hot water", "nicking", "abrasion", etc---and scarifTypeGen which was just "chemical" , "mechanical", "moisture."

> 2. I've tried my best to fix any mistakes in the species names that stood out as obviously wrong, but I don't know all the species names in the world and so was wondering how to best correct any other mistakes I may have missed. For example, Rubus idaeus was sometimes spelled Rubus idaeust with the "t". I remember for EGRET we compared against a plant database (from Kew? Plants of the World?) so if anyone could give some pointers as to how to approach this, please let me know!

Yes, there are R packages for this. For an example, you can check out the code @wangxm-forest wrote using the taxize package [here](https://github.com/lizzieinvancouver/egret/blob/main/analyses/cleaning/source/cleanspecies.R).

**Comment by ngoj1** [comment]
Thank you for such a fast reply! I've implemented those changes now with the exception of the pretreatmentFeces column just showing which type of feces were used, as there's already a seed_type column that indicates fresh/dry/etc. seeds. I also gave "heat" its own ScarifTypeGen as "Thermal". 

All species names have now passed the check except for the raspberry cultivar 'Glen Cova' but that's just the exception because it's a cultivar and not a true species or subspecies/variety/form.

There's just one last thing I need to check which is Rubus spectabilis which has no details on scarification type, but after that then the germination data is all good to go.

**Comment by selenashew** [comment]
June 24 Updates:

1. The germination dataset is ready to go!! The finalized, clean germination data can be found under 'egret\analyses\scrapeUSDAseedmanual\output\germinationCleaned_official.xlsx'. Justin and I want to note that after running the data through Justin's cleaning script, we did one more manual check of the data and fixed several weird, one-off cases (eg. the PDF lists day & night temps as åI and å7 for Prunus cerasifera; I found the original articles that listed the temps as -1 and -7 and changed that accordingly).

2. Britany's data preparation script successfully worked for the phenology data and she will be working on retooling it for the seedcrop data as well! I will now be working on merging/cleaning the phenology data.

**Comment by lizzieinvancouver** [comment]
@selenashew That sounds like great news! Thank you and @ngoj1 and @buniwuuu for all your work! We'll take a look at the output soon and let you know if we have any questions or see any final things to tweak.

**Comment by DeirdreLoughnan** [comment]
@selenashew @buniwuuu @ngoj1 thanks for working on this.

I took a quick glance through the germinationCleaned_original file. But here are a few things that jumped out that should be fixed:

- [ ] Can we make column names in the USDA data the same as those we have in the egret data? 
- [ ] Change the #NA so it reads properly into R 
- [ ] merge all the response variables (percent_germ, avg_germination, etc) into one column like we have in egre.
- [ ] Some of the stratification_temp_C get converted to dates when the file is converted to a csv, could you change the excel to not auto-format?
- [ ] We does "Var." mean in the cold_stratification_days

@lizzieinvancouver a lot of the values are given as ranges (test_duration_in_days, total_germination_percent, day_temp_celsius for example), but written in words ("150 to 210", "67 to 72" etc). Ideally this should be continuous data. I think we should calculate the mean or extract the minimums in new columns for these variable, but we could also wait until we know exactly what data we are going to model first.

**Comment by lizzieinvancouver** [comment]
> @lizzieinvancouver a lot of the values are given as ranges (test_duration_in_days, total_germination_percent, day_temp_celsius for example), but written in words ("150 to 210", "67 to 72" etc). Ideally this should be continuous data. I think we should calculate the mean or extract the minimums in new columns for these variable, but we could also wait until we know exactly what data we are going to model first.

@DeirdreLoughnan I think it would be good to calculate the mean now into a new column as that would be a good way to check the formatting is consistent. Maybe @selenashew  or @ngoj1 or @buniwuuu could take this on. It should be done with grep in R -- getting all numbers before ` to ` and then all numbers after ` to ` ideally (unless someone has a better idea).

**Comment by selenashew** [comment]
@DeirdreLoughnan most of the genuses have values across several different response variables (eg. Malus has separate values for both Germinative Energy and Germinative Capacity). For your third point, which response variable would you like to prioritize in terms of merging and which ones should we ignore?

**Comment by ngoj1** [comment]
Selena, Britany, and I will go over all five of those points tomorrow morning and have it done ASAP; I'm estimating by Thursday midnight if we're fast, since it doesn't seem like it involves anything we haven't encountered before in our other data cleaning experiences.

We could try copying the data into a new column and then strsplit() by " to " to make them into min and max, then calculate mean?

**Comment by selenashew** [comment]
I tried looking through all of the listed sources for Cercis to determine the "Var" value for cold_stratification_days but was unfortunately unsuccessful. The PDF has no further explanation, and the closest mention I could find was from Afanasiev (1944) (https://books.google.ca/books?hl=en&lr=&id=jxlPplOGLJgC&oi=fnd&pg=PA405&dq=Afanasiev+M.+1944.+A+study+of+dormancy+and+germination+of+seeds+of+Cercis+canadensis.+Journal+of+Agricultural+Research+69(10):+405%E2%80%93420&ots=99U81_W8cF&sig=mA2Ufrw8cnOfqotwX1ul4OfdQf0#v=onepage&q&f=false) where they state that after 15 months of cold storage, some of their seeds responded better to 45 minutes of H2SO4 compared to the typical 30 minutes.

**Comment by selenashew** [comment]
Hi @DeirdreLoughnan @lizzieinvancouver,

We're currently almost done making all of the new fixes but we have a new question. We've noticed that there are a few '<' and '>' signs in the numerical values (eg. <16 hours for photoperiod). Would you like for us to keep these signs or remove them?

**Comment by lizzieinvancouver** [comment]
@selenashew If that is how the data looked in the table, we should keep them in. I assume they read into R okay? We'll likely clean them later for analysis but we try to keep the data as received/shown whenever possible. Thanks again for all your work on this!

**Comment by selenashew** [comment]
@lizzieinvancouver Sounds good! We'll keep them that way. We also noticed one more weird issue- for Rubus idaeus, their total germination is given as 3 different numbers, eg. 69 (18-94). From what we can see, the number outside of the range comes from one experiment/article and the given range comes from other experiments/articles, as there is a great amount of variability in germination for this species. Would you like for us to only use the first number (eg. 69) and remove the range (18-94), or would you prefer for us to still keep everything and calculate a new mean value for the given range?

**Comment by lizzieinvancouver** [comment]
> We also noticed one more weird issue- for Rubus idaeus, their total germination is given as 3 different numbers, eg. 69 (18-94). From what we can see, the number outside of the range comes from one experiment/article and the given range comes from other experiments/articles, as there is a great amount of variability in germination for this species. Would you like for us to only use the first number (eg. 69) and remove the range (18-94), or would you prefer for us to still keep everything and calculate a new mean value for the given range?

@selenashew Ugh, that is a hard one. I think if the authors have given a mean, then we should use it (and not calculate a new one). So we should leave the value 69 (18-94) in the original cell, but enter 69 in the mean cell and add notes in the code for why (reference this git issue also). 

Thank you again!

**Comment by ngoj1** [comment]
> So we should leave the value 69 (18-94) in the original cell, but enter 69 in the mean cell and add notes in the code for why (reference this git issue also).

Just added a comment to the new code that references this issue! All the columns that had ranges have now had new columns made for the Min and Max with a third column to represent the mean. The three of us will meet up on Teams tomorrow to look over the code and the output to make sure everything looks good.

**Comment by DeirdreLoughnan** [comment]
@ngoj1 @buniwuuu @selenashew Amazing thank you all for your great work!

**Comment by ngoj1** [comment]
@lizzieinvancouver @DeirdreLoughnan Selena, Britany, and I are practically done with the USDA table cleaning. Germination data should be good to go, mean average columns are done, column names changed, etc. Selena just has to go over the column names in the Phenology and Seed Data tables (which have been combined) in the original pdf to see what they mean because some of them are a bit confusing like "fruit colour ripeness criteria", but after that all should be good! Those tables will all be double checked and submitted by the end of this week if that's okay.

**Comment by lizzieinvancouver** [comment]
> Germination data should be good to go, mean average columns are done, column names changed, etc. Selena just has to go over the column names in the Phenology and Seed Data tables (which have been combined) in the original pdf to see what they mean because some of them are a bit confusing like "fruit colour ripeness criteria", but after that all should be good! Those tables will all be double checked and submitted by the end of this week if that's okay.

@ngoj1 Great! Thank you and @selenashew and @buniwuuu ! We can definitely wait on the other columns, but will dive in on the germination data.

**Comment by DeirdreLoughnan** [comment]
@ngoj1 could you point me to the new file and where there is meta-data explaining what the different columns are?

I was looking at [USDAgerminationCleanedFinal.csv](https://github.com/lizzieinvancouver/egret/blob/main/analyses/scrapeUSDAseedmanual/output/USDAgerminationCleanedFinal.csv) and saw a few more cleaning tasks that we would love your help with @ngoj1 @buniwuuu @selenashew:

- [ ] Could you double check the column names? Quite a few columns still do not line up with ones in egret, like "seedSource", which I imagine should be "source.population."
- [ ] Could you cross reference the [egret responseVar cleaning code](https://github.com/lizzieinvancouver/egret/blob/main/analyses/cleaning/source/cleanResponseVar.R) and rename the responseType's to align with what I did for the egret data?

**Comment by DeirdreLoughnan** [comment]
@lizzieinvancouver here is a quick summary of the USDA data:

- It includes 310 species from 39 genus
- 945 rows of data for five different response variables, all of which overlap with the egret data, although we might care less about germination capacity and germination energy.
- We have 208 rows of percent germination data for 134 species, just over half (132 rows) have temperatures data, which I think it is chilling since the values range from -1 to 8.  Of this 66% includes information on the duration, so we could calculate chilling.  As far as I can tell there is no germination temperature data. 
- There is 75 rows of data on germination rate, for 51 species. The temperature data for this response variable spans a larger range.

Some of this might become clearer once I find the meta-data and have a better idea what some of the columns mean though.

**Comment by ngoj1** [comment]
> could you point me to the new file and where there is meta-data explaining what the different columns are?

I'll double check with the paper to see what these column names are so that I can generate a metadata file. Since Selena will be busy packing this weekend, Britany and I can handle the those two tasks!

**Comment by ngoj1** [comment]
Ok, here is an update:
 - Metadata can be found in "egret/analyses/scrapeUSDAseedmanual/output/**usdaGerminationMetadata.csv**"
 - Cleaned data can be found in "egret/analyses/scrapeUSDAseedmanual/output/**usdaGerminationData.csv**"
 - Both of these can be found as sheets in the Excel file "egret/analyses/scrapeUSDAseedmanual/output/**usdaGerminationMaster.xlsx**"

I'll get started on making them more like the EGRET datasheet now!

**Comment by ngoj1** [comment]
> Could you double check the column names? Quite a few columns still do not line up with ones in egret, like "seedSource", which I imagine should be "source.population."

I tried to match columns where I felt like they were probably expressing the same thing ("pretreatmentChillDuration" became "chill.duration" for example) but there were also some that were unique to this dataset so I left them in camel case. I changed "responseType" to "responseVar" and not "response.value" because I saw there was a "responseVar" in the EGRET column headers, but I can change any of these things to a different format.

> Could you cross reference the [egret responseVar cleaning code](https://github.com/lizzieinvancouver/egret/blob/main/analyses/cleaning/source/cleanResponseVar.R) and rename the responseType's to align with what I did for the egret data?

Anything that matched what was in EGRET was changed to fit it, like germinationTime becoming "mgt" since we now have a responseValueAvg column. There were some response variables that I had trouble matching, but I kept the format of using the period in between words, e.g. "germCapacityAvg" became "mean.germ.capacity" to mirror the "median.germ.time" in EGRET. If you would prefer it to be "m.germ.capacity" to match "mgt" I can quickly change that, I just left it unabbreviated because I thought it was more explicit.

Would you also like me to remove any rows in which there isn't a responseValue entered?

**Comment by lizzieinvancouver** [comment]
As this is the issue for the first BIG round of cleaning and that is DONE (yay! nice work @ngoj1 @selenashew @buniwuuu and thank you!) I am closing this and refer onward to issues #36 and #53 .
